msdm/algorithms/fscgradientascent.py

```diff
@@ -0,0 +1,161 @@
+    V = V.view((ncontroller, nstates))
+
+    # This is a departure from Meuleau 1999; their initial controller state
+    # distribution is conditional on an initial observation. However, our
+    # formalization of the POMDP has an observation function that's conditional
+    # on action. We can surely implement their initial controller state distribution
+    # for some classes of POMDP, but leave that out for now.
+    state_value = fsc_initial_state @ V
+
+    return Result(
+        state_controller_value=V,
+        state_value=state_value,
+        expected_value=state_value@s0,
+    )
+
+class FSCGradientAscent(Learns):
+    def __init__(
+        self, *,
+        # Number of states the controller should have.
+        controller_state_count,
+        iterations=5000,
+        learning_rate=1e-1,
+        # This parameter is the number of iterations that should pass between updates printed  out
+        log_iteration_progress=None,
+        # Threshold used to detect convergence
+        convergence_diff=1e-4,
+        optimizer=torch.optim.Adam,
+        dtype=torch.float64,
+        seed=None,
+    ):
+        self.dtype = dtype
+        self.learning_rate = learning_rate
+        self.log_iteration_progress = log_iteration_progress
+        self.iterations = iterations
+        self.convergence_diff = convergence_diff
+        self.controller_state_count = controller_state_count
+        self.optimizer = optimizer
+        self.seed = seed or torch.randint(2**30, size=(1,)).item()
+
+    def train_on(self, pomdp: TabularPOMDP):
+        # Number of states, actions, observations in the POMDP
+        nactions, nstates, nobs = pomdp.observation_matrix.shape
+        # Number of states in the finite state controller.
+        ncontroller = self.controller_state_count
+
+        with torch.random.fork_rng():
+            torch.random.manual_seed(self.seed)
+            fsc_action_logit = torch.rand(ncontroller, nactions, requires_grad=True, dtype=self.dtype)
+            fsc_state_logit = torch.rand(ncontroller, nactions, nobs, ncontroller, requires_grad=True, dtype=self.dtype)
+            fsc_initial_state_logit = torch.rand(ncontroller, requires_grad=True, dtype=self.dtype)
+
+        def value():
+            return stochastic_fsc_policy_evaluation_exact(
+                pomdp,
+                fsc_action_logit.softmax(-1),
+                fsc_state_logit.softmax(-1),
+                fsc_initial_state_logit.softmax(-1),
+                dtype=self.dtype,
+            )
+
+        opt = self.optimizer([fsc_action_logit, fsc_state_logit, fsc_initial_state_logit], lr=self.learning_rate)
+
+        prevV = None
+        converged = False
+
+        for idx in range(self.iterations):
+            opt.zero_grad()
+
+            result = value()
+            V = result.state_value
+
+            # Check for convergence
+            if prevV is not None:
+                max_bellman_error = torch.abs(V-prevV).max().item()
```

Comment:
Flagging this for two reasons:
- a traditional convergence test through backprop seems really hard to pass with softmax-based distributions
- Doing this for state-wise values (i.e. after marginalizing out FSC state), but not quite sure if that's right.
Paper also doesn't mention convergence, so may be simplest to ax.